File: app/src/main/java/com/example/learn/testme/CategoryFragment.java
package com.example.learn.testme;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.learn.testme.Common.common;
import com.example.learn.testme.Interface.ItemClickLIstener;
import com.example.learn.testme.Model.Category;
import com.example.learn.testme.ViewHolder.CategoryViewHolder;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.squareup.picasso.Picasso;


public class CategoryFragment extends Fragment {

    View myfragment;

    RecyclerView listCategory;
    RecyclerView.LayoutManager layoutManager;
    FirebaseRecyclerAdapter<Category,CategoryViewHolder> adapter;

    FirebaseDatabase database;
    DatabaseReference categories;

    public static CategoryFragment newInstance(){
        CategoryFragment categoryFragment = new CategoryFragment();
        return categoryFragment;

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        database = FirebaseDatabase.getInstance();
        categories = database.getReference("Category");
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        myfragment = inflater.inflate(R.layout.fragment_category , container,false);

        listCategory = (RecyclerView)myfragment.findViewById(R.id.listCategory);
        listCategory.setHasFixedSize(true);
        layoutManager = new LinearLayoutManager(container.getContext());
        listCategory.setLayoutManager(layoutManager);

        loadCategories();
        return myfragment;
    }

    private void loadCategories() {

        adapter = new FirebaseRecyclerAdapter<Category, CategoryViewHolder>(
                Category.class,
                R.layout.category_layout,
                CategoryViewHolder.class,
                categories
        ){
            @Override
            protected void populateViewHolder(CategoryViewHolder viewHolder, final Category model, int position) {
                viewHolder.category_name.setText(model.getName());
                Log.e("caategory >> ", model.getImage());
                Picasso.with(getActivity())
                        .load(model.getImage())
                        .into(viewHolder.category_image);

                viewHolder.setItemClickListener(new ItemClickLIstener() {
                    @Override
                    public void onClick(View view, int position, boolean isLongClick) {

                        FragmentTransaction transaction = getActivity().getSupportFragmentManager().beginTransaction();
                        transaction.replace(R.id.frame_layout,Subject.newInstance());
                        transaction.commit();



                        Log.e("position >>",adapter.getRef(position).getKey());
                        common.categoryId = adapter.getRef(position).getKey();
                        common.categoryName = model.getName();



                    }
                });
            }
        };

        adapter.notifyDataSetChanged();
        listCategory.setAdapter(adapter);
    }


}

File: app/src/main/java/com/example/learn/testme/CategoryFragment.kt
package com.example.learn.testme

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.learn.testme.Model.Category
import com.example.learn.testme.ViewHolder.CategoryViewHolder
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

class CategoryFragment : androidx.fragment.app.Fragment() {
    var myfragment: View? = null
    var listCategory: androidx.recyclerview.widget.RecyclerView? = null
    var layoutManager: androidx.recyclerview.widget.RecyclerView.LayoutManager? = null
    var adapter: FirebaseRecyclerAdapter<Category?, CategoryViewHolder>? = null
    var database: FirebaseDatabase? = null
    var categories: DatabaseReference? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        database = FirebaseDatabase.getInstance()
        categories = database!!.getReference("Category")
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        myfragment = inflater.inflate(R.layout.fragment_category, container, false)
        listCategory = myfragment!!.findViewById<View>(R.id.listCategory) as androidx.recyclerview.widget.RecyclerView
        listCategory!!.setHasFixedSize(true)
        layoutManager = androidx.recyclerview.widget.LinearLayoutManager(container!!.context)
        listCategory!!.layoutManager = layoutManager
        loadCategories()
        return myfragment
    }

    private fun loadCategories() {
//        adapter = object : FirebaseRecyclerAdapter<Category, CategoryViewHolder>(
//                Category,
//                R.layout.item_category,
//                CategoryViewHolder::class.java,
//                categories
//        ) {
//            override fun populateViewHolder(viewHolder: CategoryViewHolder, model: Category, position: Int) {
//                viewHolder.category_name.text = model.name
//                Log.e("caategory >> ", model.image)
//                Picasso.with(activity)
//                        .load(model.image)
//                        .into(viewHolder.category_image)
//                viewHolder.setItemClickListener { view, position, isLongClick ->
//                    val transaction = activity!!.supportFragmentManager.beginTransaction()
//                    transaction.replace(R.id.frame_layout, Subject.newInstance())
//                    transaction.commit()
//                    Log.e("position >>", adapter!!.getRef(position).key)
//                    common.categoryId = adapter!!.getRef(position).key
//                    common.categoryName = model.name
//                }
//            }
//        }
        adapter!!.notifyDataSetChanged()
        listCategory!!.adapter = adapter
    }

    companion object {
        fun newInstance(): CategoryFragment {
            return CategoryFragment()
        }
    }
}